### ISSUE TITLE
gradient descent doesn't work on more than 2 qubits

### ISSUE THREAD

**Issue by SimonR99** [MatchCake/MatchCake#31]
In the code bellow, the the first circuit is executed, but the second one isn't. The only difference is the requires_grad.

```python
import pennylane as qml
from matchcake import NonInteractingFermionicDevice
from matchcake.operations import MAngleEmbedding, fRYY
from pennylane.wires import Wires
import torch

import warnings
warnings.filterwarnings('ignore') # TODO : Fix warnings

n_qubits = 4

x = torch.tensor([0.111, 0.012, 0.11,0.22])
x_grad = torch.tensor([0.111, 0.012, 0.11,0.22], requires_grad=True)

dev = NonInteractingFermionicDevice(wires=n_qubits)
#dev = qml.device('default.qubit', wires=n_qubits)

@qml.qnode(dev, interface="torch")
def circuit(x):
    fRYY(x[0:2], wires=[0, 1])
    fRYY(x[2:4], wires=[2, 3])
    return qml.expval(qml.Projector([0]*n_qubits, wires=range(n_qubits)))


print(circuit(x))
print(circuit(x_grad))
```

If you change the second fRYY array to be [0,1] instead of [2,3], everything works fine.

**Comment by JeremieGince** [comment]
The problem appears to be related to the contraction method, so I added the test 'test_nh_contraction_torch_grad' in 'test/test_nif_device/test_device_utils' to ensure that the contraction method no longer raises an error with the provided code snippet.
